src/lib/provable/gadgets/sha2.ts

```diff
@@ -0,0 +1,621 @@
+    // padding the message $5.1.1 into blocks that are a multiple of 512
+    let messageBlocks = padding<Type>(length, data);
+
+    let H = SHA2.initialState<Type>(length);
+
+    const N = messageBlocks.length;
+
+    for (let i = 0; i < N; i++) {
+      const W = messageSchedule(length, messageBlocks[i]);
+      H = compression(length, H, W);
+    }
+
+    // the working variables H[i] are 32 | 64 bit, however we want to decompose
+    // them into bytes to be more compatible
+    let digest = Bytes.from(H.map((x) => x.toBytesBE()).flat());
+
+    // Take the first `length` bits of the digest. This has no effect in
+    // SHA256 and SHA512, because 8 words of 32 | 64 bits needs no truncation.
+    // Truncation is required for SHA224 and SHA384 though.
+    digest.bytes = digest.bytes.slice(0, length / 8);
+
+    return digest;
+  },
+  compression,
+  messageSchedule,
+  padding,
+  initialState<T extends UInt32 | UInt64>(length: Length): T[] {
+    switch (length) {
+      case 224:
+      case 256:
+        return SHA2Constants.H[length].map((x) => UInt32.from(x) as T);
+      case 384:
+      case 512:
+        return SHA2Constants.H[length].map((x) => UInt64.from(x) as T);
+    }
+  },
+};
+
+/**
+ * Padding function for SHA2, as specified in §5.1.1, §5.1.2,
+ *
+ * @param data  The message to hash
+ * @param length Whether this is a SHA2-224 or SHA2-256 or SHA2-384 or SHA2-512
+ * @returns
+ */
+// The only difference between the padding used in SHA2-224/256 and SHA2-384/512
+// is the size of the word (32bit vs 64bit). In the first case, UInt32[][] is
+// returned, in the second case UInt64[][] is returned.
+function padding<T extends UInt32 | UInt64>(
+  length: Length,
+  data: FlexibleBytes
+): T[][] {
+  // create a provable Bytes instance from the input data
+  // the Bytes class will be static sized according to the length of the input data
+  let message = Bytes.from(data);
+
+  const blockLength = SHA2Constants.BLOCK_LENGTH[length];
+  const paddingValue = SHA2Constants.PADDING_VALUE[length];
+  const lengthChunk = SHA2Constants.LENGTH_CHUNK[length];
+
+  // now pad the data to reach the format expected by SHA2
+  // pad 1 bit, followed by k zero bits where k is the smallest non-negative solution to
+  // l + 1 + k = (448 | 896) mod (512 | 1024)
+  // then append a (64 | 128)-bit block containing the length of the original message in bits
+  // it holds that PADDING_VALUE = BLOCK_LENGTH - LENGTH_CHUNK
+  // this way the padded message will be a multiple of the BLOCK_LENGTH
+
+  let l = message.length * 8; // length in bits
+  let k = Number(mod(paddingValue - (BigInt(l) + 1n), blockLength));
+
+  let lBinary = l.toString(2);
+
+  let paddingBits = (
+    '1' + // append 1 bit
+    '0'.repeat(k) + // append k zero bits
+    '0'.repeat(lengthChunk - lBinary.length) + // append 64|128-bit containing the length of the original message
+    lBinary
+  ).match(/.{1,8}/g)!; // this should always be divisible by 8
```

Comment:
Oh. Is it a common practice in TS? :o

Reply:
Not common, but it's something I had to do in the other SHA implementation a while ago - cant remember why tho